ev3dev2/led.py

```diff
@@ -346,3 +385,231 @@ def all_off(self):
+
+    def animate_stop(self):
+        """
+        Signal the current animation thread to exit and wait for it to exit
+        """
+
+        if self.animate_thread_id:
+            self.animate_thread_stop = True
+
+            while self.animate_thread_id:
+                pass
+
+    def animate_police_lights(self, color1, color2, group1='LEFT', group2='RIGHT', sleeptime=0.5, duration=5, block=True):
+        """
+        Cycle the ``group1`` and ``group2`` LEDs between ``color1`` and ``color2``
+        to give the effect of police lights.  Alternate the ``group1`` and ``group2``
+        LEDs every ``sleeptime`` seconds.
+
+        Animate for ``duration`` seconds.  If ``duration`` is None animate for forever.
+
+        Example:
+
+        .. code-block:: python
+
+            from ev3dev2.led import Leds
+            leds = Leds()
+            leds.animate_police_lights('RED', 'GREEN')
+        """
+
+        def _animate_police_lights():
+            self.all_off()
+            even = True
+            start_time = dt.datetime.now()
+
+            while True:
+                if even:
+                    self.set_color(group1, color1)
+                    self.set_color(group2, color2)
+                else:
+                    self.set_color(group1, color2)
+                    self.set_color(group2, color1)
+
+                if self.animate_thread_stop or duration_expired(start_time, duration):
+                    break
+
+                even = not even
+                sleep(sleeptime)
+
+            self.animate_thread_stop = False
+            self.animate_thread_id = None
+
+        self.animate_stop()
+
+        if block:
+            _animate_police_lights()
+        else:
+            self.animate_thread_id = _thread.start_new_thread(_animate_police_lights, ())
+
+    def animate_flash(self, color, groups=('LEFT', 'RIGHT'), sleeptime=0.5, duration=5, block=True):
+        """
+        Turn all LEDs in ``groups`` off/on to ``color`` every ``sleeptime`` seconds
+
+        Animate for ``duration`` seconds.  If ``duration`` is None animate for forever.
+
+        Example:
+
+        .. code-block:: python
+
+            from ev3dev2.led import Leds
+            leds = Leds()
+            leds.animate_flash('AMBER')
+        """
+
+        def _animate_flash():
+            even = True
+            start_time = dt.datetime.now()
+
+            while True:
+                if even:
+                    for group in groups:
+                        self.set_color(group, color)
+                else:
+                    self.all_off()
+
+                if self.animate_thread_stop or duration_expired(start_time, duration):
+                    break
+
+                even = not even
+                sleep(sleeptime)
+
+            self.animate_thread_stop = False
+            self.animate_thread_id = None
+
+        self.animate_stop()
+
+        if block:
+            _animate_flash()
```

Comment:
Do we want to break this common logic into a method like `animate(do_animate, ...)` so it's easy to extend with custom animations? (`do_animate` would be a function). It would be really nice if we could find a way to include the boilerplate logic within the thread (e.g. stopping after time or when cancelled) "for free" but I'm not sure how would be best. Maybe the lambda just runs one LED update step and tracks its own state?

Reply:
I am not a big fan of ``lambda`` myself...I find the resulting code much harder to read/follow :( There is a some repetitive logic in these animate methods but not a crazy amount.  I would like to leave it as-is.